### ISSUE TITLE
vscode + rust analyzer close rustc processes on exit

### ISSUE THREAD

**Issue by jordy25519** [rust-lang/rust-analyzer#5258]
Quitting vscode leaves some running rustc/cargo jobs holding a lock on the build directory.
Requires a manual `kill` to stop them.

Environment:
```
Version: 1.46.1
Commit: cd9ea6488829f560dc949a8b2fb789f3cdc05f5d
Date: 2020-06-17T21:17:14.222Z
Electron: 7.3.1
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Darwin x64 17.7.0
```
```
Name: rust-analyzer
Id: matklad.rust-analyzer
Description: An alternative rust language server to the RLS
Version: 0.2.232
```

```
> ~/Library/Application\ Support/Code/User/globalStorage/matklad.rust-analyzer/rust-analyzer-mac --version
rust-analyzer 8b0983e
```

**Comment by bjorn3** [comment]
You have to be patient for rustc to finish compiling.

**Comment by matklad** [comment]
We do kill cargo on death: https://github.com/rust-analyzer/rust-analyzer/blob/41feb816c98894d48df97193eb085738e405a3c2/crates/flycheck/src/lib.rs#L312-L316

And Cargo has some special magic to clean the whole process-group. So, not sure what's going on here, needs investigation! But I think I've seen this myself. 

The expected behavior is that we promptly cleanup all processes on normal shutdown (it's OK to leave processes hanging if we are sigkilled ourselves, as properly setting up cross-platform process groups is a pain).

**Comment by oeed** [comment]
This one definitely happens for me a lot. Not sure if it's something about my setup, but I frequently have to kill cargo processes before rust-analyzer starts working in VSCode; it just hangs waiting for the file lock otherwise. Would be great to resolve.

**Comment by jak-pan** [comment]
Same here. Happens a lot. Getting crazy from this. Happy to debug.

**Comment by oeed** [comment]
@lnicola could you possibly explain the unactionable tag?

**Comment by lnicola** [comment]
@oeed I marked it as unactionable because there's no way to reproduce it, and the code, as written, shouldn't leak `cargo` or `rustc` processes, at least on Unix platforms. Even if a compiler is left running, it should finish its job then close, just as if you had ran it from the terminal.

Personally, I've never seen this happen on Linux or Windows.

**Comment by MikailBag** [comment]
Looking at cargo source documentation, I see this (`cargo::utils::job`):

>  This currently works "by default" on Unix platforms because Ctrl-C actually sends a signal to the process group rather than the parent process, so everything will get torn down.

`man 2 kill` tells:
> If pid is positive, then signal sig is sent to the process with the ID specified by pid. 
If pid equals 0, then sig is sent to every process in the process group of the calling process.
If pid equals -1, then sig is sent to every process for which the calling process has permission to send signals, except for process 1 (init), but see below. If pid is less than -1, then sig is sent to every process in the process group whose ID is -pid.

Rust-analyzer uses `Child::kill` method to kill `cargo` process. [Here](https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/std/sys/unix/process/process_inner/struct.Process.html#method.kill) is Unix implementation. You can see that positive value is being passed to `kill` syscall, which means `rustc`s will not receive SIGKILL and continue running.

**Comment by matklad** [comment]
Oh, right, if we just `kill` cargo, it wont necessary kill the processes spawned by the cargo.

@oconnor663, can I tap into your knowledge a bit here? You are the expert on the topic: https://github.com/oconnor663/duct.py/blob/master/gotchas.md#killing-grandchild-processes :-)

So the situation is that rust-analyzer is not spawned by the shell, it's spawned by the editor (GUI vscode, teminal VIM, etc). rust-analyzer itself spawns cargo, and cargo spawns rustc. 

The shutdown process for rust-analyzer is cooperative -- the client sends us a "shutdown" message, we cancel in-progress jobs, wait for them to finish, and then exit outselves.

The problem is that one of the in-progress jobs might be `cargo check`. At the moment, we just kill Cargo. Re-reading https://github.com/rust-lang/cargo/blob/1243c378b9a637dc4beb9b2ad7b955af3affa47e/src/cargo/util/job.rs#L1-L16, I am not sure if that's enough -- rust-analyzer is not a shell, so we don't do anything specific about process groups. 

Is there anything better we can do here? Should rust-analyzer play the role of the shell and try to setup process group here? Or should we escalate this further up and say that whoever spawned rust-analyzer should also clean dependent processes? That is, that *they* should setup a process group? Should we just :shrug: and mark this as a won't fix?

**Comment by jak-pan** [comment]
Don't know if it helps your decision but you can specify different process to run in vscode settings (it's cargo check by default but running cargo clippy is also possible)

**Comment by MikailBag** [comment]
IMHO the simplest solution is to use `pid namespaces`. Although this is a Linux-only feature, I think it should be a good fit for this case.
It should be used in such a way that the rust-analyzer main process is `init` process inside the namespace. That way when the rust-analyzer main process exits all its children (including proc-macro server, cargo, rustc, etc) are killed by the kernel automatically.

**Comment by lnicola** [comment]
Windows also has [Job Objects](https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects) for this. But it's not clear to me why this is actually a problem. Doesn't `rustc` finish by itself in a couple of seconds? Does it hang if `cargo` died?

**Comment by MikailBag** [comment]
Well, I think I suffer from this problem when my laptop's RAM is overloaded with firefox, vscode, rust-analyzer, and compilation. If one of the rustc processes links binary, it can take quite a lot of time (because of swapping). I haven't measured but sometimes rustc keeps running even after a few seconds.

**Comment by lnicola** [comment]
Ah, a few seconds should be fine. I interpreted the first post as "the remaining processes never close until killed".

**Comment by jak-pan** [comment]
One more note from me. I use format on save in vscode which runs rustfmt. It happens when cargo check or formatter is running and I press save again, not sure exactly at which point. It looks like the processess interlock each other and never finish. Is it possible that they both try to file lock and stay running indefinitly because of that?

Looks like when I save the old cargo process is not killed

I can easily repeat by loading large project and smashing cmd+s multiple times on different osx machines.

![image](https://user-images.githubusercontent.com/3409250/115235183-31b37c00-a11a-11eb-8050-ac0320194afd.png)

**Comment by MikailBag** [comment]
> Ah, a few seconds should be fine

I have never measured how long these `ld` instances can run for if I don't interrupt them. And it seems the situation gets worse for users with less RAM or slower HDD.

**Comment by jak-pan** [comment]
Also this is happening when this happens.
<img width="572" alt="Screenshot 2021-04-19 at 14 20 14" src="https://user-images.githubusercontent.com/3409250/115235441-793a0800-a11a-11eb-9abb-f47a7452acab.png">

**Comment by lnicola** [comment]
I couldn't reproduce (on Linux) in the RA repo, neither by saving 50 times in quick succession, not by holding Ctrl-S pressed for about 10 seconds.

**Comment by oconnor663** [comment]
Not sure I'm gonna add anything here that you all haven't already thought of, but just riffing: One fix could be for Cargo to cooperate also, by handling SIGINT and SIGTERM and explicitly propagating those signals to its children. This doesn't work for SIGKILL, so rust-analyzer would probably need to switch to sending SIGTERM. I don't know if `TerminateProcess` can be handled on Windows, though, so this solution might be Unix-only. (But maybe that's tolerable since we're happy to use the Jobs API on Windows?) This also assumes Cargo doesn't have any long-lived grandchildren, which it certainly might via `build.rs`. (Are there other examples?)

If we ask the caller to put rust-analyzer in its own process group, and then have the caller signal the group, that could work. But I assume that's a big API change? And are there other reasons to want to stick with cooperative shutdown?

If rust-analyzer used process groups internally, that could make it possible to reliably cancel whole jobs without shutting down rust-analyzer itself. Is that desirable? (Is it possible that rust-analyzer already thinks it can do this, but actually isn't shutting down grandchildren promptly? What happens to the Cargo lockfile if Cargo is killed but rustc grandchildren are still running?)

The biggest downside with process groups in general is that each layer of the stack breaks the layers above it. So if you spawned VSCode in the terminal, and you want Ctrl-C in the terminal to kill the whole thing, any process group ninjutsu in or around rust-analyzer will mean that Ctrl-C doesn't reach it. Similarly, any future decision to use process groups _inside_ Cargo could break rust-analyzer's strategy.

So...yeah...this stuff makes me sad. If rust-analyzer already has a cooperative shutdown API, maybe that's an ok place to "draw a line" for signaling purposes? Also since it can shut itself down if it detects the parent has closed its write pipe?

Incidentally the [code you linked](https://github.com/rust-lang/cargo/blob/1243c378b9a637dc4beb9b2ad7b955af3affa47e/src/cargo/util/job.rs#L35) looks like it's calling `std::env::var` (which allocates) inside of a post-fork-pre-exec function. That's probably unsound :(

**Comment by GopherJ** [comment]
Same here on mbp m1, it's not a thing of `seconds` but can be a thing of `30 seconds` even `minutes` while working on substrate & polkadot projects:)

**Comment by holygits** [comment]
for this building on substrate, these settings have helped to lessen the problem for me
```
    "rust-analyzer.cargo.runBuildScripts": false, <-- don't run the wasm build every time
    "rust-analyzer.checkOnSave.allTargets": false, <-- don't run test build every time
```

**Comment by destenson** [comment]
> @oeed I marked it as unactionable because there's no way to reproduce it, and the code, as written, shouldn't leak `cargo` or `rustc` processes, at least on Unix platforms. Even if a compiler is left running, it should finish its job then close, just as if you had ran it from the terminal.
> 
> Personally, I've never seen this happen on Linux or Windows.

I'm currently dealing with this problem, or a similar one. I have my editor set to autosave quite often (but it happens when I save often manually as well). It seems that each time a file is saved, cargo is run one or more times. When the first run isn't completed by the time the second is started, it appears to sit & wait for the lock file to be gone to proceed. When saving often, this causes sometimes dozens of cargo processes to run.

As others have noted, this may occur more often when RAM is full and a lot of swapping is going on. I am seeing this behavior on both Linux & Windows.

**Comment by lnicola** [comment]
@destenson do they all finish if you stop editing the code?

**Comment by destenson** [comment]
> @destenson do they all finish if you stop editing the code?

If there aren't too many, they do. But I have to actively monitor processes & limit how often I save in order to prevent too many from building up. They _are_ all killed when I exit code, so I guess this isn't the exact same bug as this thread is about.

**Comment by lnicola** [comment]
Yeah, we don't stop running `cargo check`s on save, that's a known issue. My suggestion would be to disable auto-saving. We have native diagnostics for syntax errors (and not only), you can configure semantic highlighting to mark unresolved symbols (so you know when you're calling the wrong function) and for anything else, a Ctrl/Cmd-S every once in a while should be fine.

But if `cargo check` takes too long, it might be rebuilding more than the current crate. That can happen if you have some mismatched environment variables, but I've also seen it happen for no good reason, and a `cargo clean` fixed it.

**Comment by jak-pan** [comment]
This happens specifically with large projects that take a bit of time to finish like paritytech/polkadot and I think it has something to do with locking the directories. It seems like in some specific case / timing you can interlock two cargo checks as killing one of the processes can unlock the other one and let it finish. Otherwise they sit at 0%. CPU. Killing the first process would certainly help in that case.

**Comment by aewc** [comment]
> You have to be patient for rustc to finish compiling.

but it's too stupid, such as there is some else error(network, permission, and so on), if I run `cargo build`, I will get where the error happens, but the analyzer always repeat to load workspace
<img width="543" alt="image" src="https://user-images.githubusercontent.com/103259156/165883916-62602114-8b52-4943-a598-cb169de12861.png">

**Comment by lnicola** [comment]
@Veykril I think this was fixed in #12660? I checked and the initial flycheck gets stopped when closing the client.
